src/NHibernate/Linq/NhRelinqQueryParser.cs

```diff
@@ -26,9 +27,11 @@ static NhRelinqQueryParser()
 			transformerRegistry.Register(new RemoveRedundantCast());
 			transformerRegistry.Register(new SimplifyCompareTransformer());
 
-			var processor = ExpressionTreeParser.CreateDefaultProcessor(transformerRegistry);
```

Comment:
This was creating a `CompoundExpressionTreeProcessor` containing a `PartialEvaluatingExpressionTreeProcessor` then a `TransformingExpressionTreeProcessor`.  
But this parser is currently only used by `NhLinqExpression`, which has already processed the expression with a `PartialEvaluatingExpressionTreeProcessor`. This second evaluation was sometimes able of evaluating further some sub-trees, altering constants which were awaiting conversion to query parameters, and causing them to no more be converted to parameters. This was occurring with `MappedAs` on a non nullable constant compared to a nullable entity property. See [NH-3961](https://nhibernate.jira.com/browse/NH-3961) comments for more details.

(Unfortunately, this change may be a possible breaking change for some external library, since `NhRelinqQueryParser` is public and could have other usages outside of NHibernate.)